src/searching/saddleback_search.rs

```diff
@@ -1,85 +1,196 @@
-// Saddleback search is a technique used to find an element in a sorted 2D matrix in O(m + n) time,
-// where m is the number of rows, and n is the number of columns. It works by starting from the
-// top-right corner of the matrix and moving left or down based on the comparison of the current
-// element with the target element.
 use std::cmp::Ordering;
 
-pub fn saddleback_search(matrix: &[Vec<i32>], element: i32) -> (usize, usize) {
-    // Initialize left and right indices
+/// Custom error type to represent errors related to matrix validation.
+#[derive(Debug, PartialEq, Eq)]
+pub enum MatrixError {
+    NonRectangularMatrix,
+}
+
```

Comment:
I propose the `is_sorted` function as follow, do you have any suggestion, @vil02?

```suggestion

/// Checks if the given matrix (vector of vectors) is sorted row-wise and column-wise.
///
/// A matrix is considered sorted if:
///
/// * Each row is sorted in non-decreasing order.
/// * Each column is sorted in non-decreasing order.
///
/// # Arguments
///
/// * `matrix` - A vector of vectors representing the matrix to check.
///
/// # Returns
///
/// Returns `true` if the matrix is sorted both row-wise and column-wise. Otherwise, returns `false`.
pub fn is_sorted(matrix: &[Vec<isize>]) -> bool {
    if matrix.is_empty() || matrix.iter().all(|row| row.is_empty()) {
        return true;
    }

    let rows = matrix.len();
    let cols = matrix[0].len();

    // Check if rows are sorted.
    for row in matrix {
        if row.windows(2).any(|w| w[0] > w[1]) {
            return false;
        }
    }

    // Check if columns are sorted.
    for col in 0..cols {
        for row in 1..rows {
            if matrix[row - 1][col] > matrix[row][col] {
                return false;
            }
        }
    }

    true
}

```